controllers/atlasmigration_controller.go

```diff
@@ -16,39 +16,310 @@ limitations under the License.
+type MigrateCLI interface {
+	Apply(ctx context.Context, data *atlas.ApplyParams) (*atlas.ApplyReport, error)
+}
+
 // AtlasMigrationReconciler reconciles a AtlasMigration object
 type AtlasMigrationReconciler struct {
 	client.Client
+	CLI    MigrateCLI
 	Scheme *runtime.Scheme
 }
 
+// HCLTmplData is the data used to render the HCL template
+// that will be used for Atlas CLI
+type HCLTmplData struct {
+	URL       string
+	Migration struct {
+		Dir string
+	}
+	Cloud struct {
+		URL     string
+		Token   string
+		Project string
+	}
+	Data struct {
+		RemoteDir struct {
+			Name string
+			Tag  string
+		}
+	}
+}
+
 //+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
 func (r *AtlasMigrationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+
+	var (
+		am  dbv1alpha1.AtlasMigration
+		err error
+	)
+
+	// Get AtlasMigration resource
+	if err = r.Get(ctx, req.NamespacedName, &am); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 
-	// TODO(user): your logic here
+	// At the end of reconcile, update the status of the resource base on the error
+	defer func() {
+		if err != nil {
+			log.Error(err, "failed to reconcile")
+		}
 
+		clientErr := r.updateResourceStatus(ctx, am, err)
+		if clientErr != nil {
+			log.Error(clientErr, "failed to update resource status")
+		}
+	}()
+
+	// Only the 'latest' version is supported
+	if am.Spec.Version != "latest" {
+		err = fmt.Errorf("unsupported version: %s", am.Spec.Version)
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile given resource
+	am.Status, err = r.reconcile(ctx, am)
+	if err != nil {
+		// Verify permanent errors to avoid infinite reconciliation loop
+		if isPernamentError(err) {
+			return result(err)
+		}
+
+		return result(transient(err))
+	}
+
+	log.Info(
+		"reconciled successfully",
+		"lastAppliedVersion", am.Status.LastAppliedVersion,
+		"lastApplied", time.Unix(am.Status.LastApplied, 0).String(),
+		"lastDeploymentUrl", am.Status.LastDeploymentURL,
+	)
 	return ctrl.Result{}, nil
 }
 
+// Reconcile the given AtlasMigration resource.
+func (r *AtlasMigrationReconciler) reconcile(
+	ctx context.Context,
+	am dbv1alpha1.AtlasMigration) (dbv1alpha1.AtlasMigrationStatus, error) {
+
+	// Build HCL template data
+	hclTmplData, cleanUpDir, err := r.buildHCLTmplData(ctx, am)
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
+	}
+	defer cleanUpDir()
+
+	// Create atlas.hcl from template data
+	file, cleanUpFile, err := executeHCLTemplate("migrateconf.tmpl", hclTmplData)
```

Comment:
We can attach the render function to data struct. So it will be
```suggestion
	atlasHCL, err := hclTmplData.render()
```

Then you can create temporary file with it, or test the render more easier